### PR TITLE
Kevmoo working

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

File: lib/src/asset/file_based.dart [hunk 1]
```diff
@@ -20,7 +20,7 @@ import 'exceptions.dart';
 /// files from disk.
 class FileBasedAssetReader implements AssetReader {
   final PackageGraph packageGraph;
-  final ignoredDirs;
+  final List ignoredDirs;
```
Comment (jakemac53) [inline lib/src/asset/file_based.dart]: might as well do `List<String>`

File: lib/src/asset/id.dart [hunk 2]
```diff
@@ -86,11 +89,12 @@ class AssetId implements Comparable<AssetId> {
   AssetId changeExtension(String newExtension) =>
       new AssetId(package, pathos.withoutExtension(path) + newExtension);
 
+  @override
   String toString() => "$package|$path";
 
   /// Serializes this [AssetId] to an object that can be sent across isolates
   /// and passed to [deserialize].
-  serialize() => [package, path];
+  List serialize() => [package, path];
```
Comment (jakemac53) [inline lib/src/asset/id.dart]: This was actually intentionally dynamic, but you could mark that explicitly (maybe as `Object` actually?). The format is likely to change in the future and is meant to be treated as opaque.

File: lib/src/asset_graph/node.dart [hunk 3]
```diff
@@ -11,7 +11,7 @@ class AssetNode {
   final AssetId id;
 
   /// The [AssetId]s of all generated assets which depend on this node.
-  final outputs = new Set<AssetId>();
+  final Set<AssetId> outputs = new Set<AssetId>();
```
Comment (jakemac53) [inline lib/src/asset_graph/node.dart]: type shouldn't be necessary here?

File: lib/src/util/constants.dart [hunk 4]
```diff
@@ -6,10 +6,10 @@ import 'dart:io';
 import 'package:crypto/crypto.dart';
 
 /// Relative path to the asset graph from the root package dir.
-final assetGraphPath = '$cacheDir/${scriptHash}/asset_graph.json';
+final String assetGraphPath = '$cacheDir/$scriptHash/asset_graph.json';
```
Comment (jakemac53) [inline lib/src/util/constants.dart]: these types shouldn't be necessary either

File: test/asset/cache_test.dart [hunk 5]
```diff
@@ -4,14 +4,13 @@
 @TestOn('vm')
 import 'dart:async';
 
-import 'package:test/test.dart';
-
 import 'package:build/build.dart';
 import 'package:build/src/asset/cache.dart';
+import 'package:test/test.dart';
```
Comment (jakemac53) [inline test/asset/cache_test.dart]: This was intentional, https://www.dartlang.org/effective-dart/style/#prefer-placing-third-party-package-imports-before-other-imports

File: test/asset/file_based_test.dart [hunk 6]
```diff
@@ -4,17 +4,17 @@
 @TestOn('vm')
 import 'dart:io';
 
+import 'package:build/build.dart';
```
Comment (jakemac53) [inline test/asset/file_based_test.dart]: ditto

File: test/asset/cache_test.dart (hunk 5 above)
Comment (jakemac53) [inline test/asset/cache_test.dart]: Same applies elsewhere as well

File: test/common/fake_watcher.dart [hunk 7]
```diff
@@ -17,18 +20,22 @@ class FakeWatcher implements DirectoryWatcher {
...
 
+  @override
   Future get ready => new Future(() {});
 
+  @override
   bool get isReady => true;
 
   /// All watchers.
-  static final watchers = <FakeWatcher>[];
+  static final List<FakeWatcher> watchers = <FakeWatcher>[];
```
Comment (jakemac53) [inline test/common/fake_watcher.dart]: same comment about the type on the left

File: test/common/matchers.dart [hunk 8]
```diff
@@ -25,7 +27,7 @@ class _AssetMatcher extends Matcher {
   const _AssetMatcher(this._expected);
 
   @override
-  bool matches(item, _) =>
+  bool matches(Object item, _) =>
```
Comment (jakemac53) [inline test/common/matchers.dart]: What value is they type here providing?

File: test/common/matchers.dart [hunk 9]
```diff
@@ -45,7 +47,7 @@ class _AssetGraphMatcher extends Matcher {
   const _AssetGraphMatcher(this._expected, this._checkValidAsOf);
 
   @override
-  bool matches(item, _) {
+  bool matches(dynamic item, _) {
```
Comment (jakemac53) [inline test/common/matchers.dart]: same comment here

**Comment by jakemac53** [comment]
Generally looks good, we should talk about the types on the left hand side for final variables, afaik there is no reason for those? Also the import ordering, I have been following the guideline to put imports for the current package below imports for other packages.

File: lib/src/asset/file_based.dart (hunk 1 above)
Comment (kevmoo) [inline lib/src/asset/file_based.dart]: ack

File: lib/src/asset/id.dart (hunk 2 above)
Comment (kevmoo) [inline lib/src/asset/id.dart]: ack

File: lib/src/util/constants.dart (hunk 4 above)
Comment (kevmoo) [inline lib/src/util/constants.dart]: In theory, no. But the lint is yelling about it.

File: lib/src/asset_graph/node.dart (hunk 3 above)
Comment (kevmoo) [inline lib/src/asset_graph/node.dart]: In theory, no. But the lint is yelling about it.

File: test/common/matchers.dart (hunk 8 above)
Comment (kevmoo) [inline test/common/matchers.dart]: Hrm – trying to make the linter happy. I really don't like this guideline at all.

File: test/common/matchers.dart (hunk 9 above)
Comment (kevmoo) [inline test/common/matchers.dart]: Ditto

**Comment by jakemac53** [comment]
So is the `type_annotate_public_apis` lint rule the one that caused types on the LHS even for final fields as well as the explicit dynamic type on arguments? In my opinion, we should remove that lint then because its providing negative value overall.

**Pull request merged**